lib/experimental/lexopt.nim

```diff
@@ -0,0 +1,251 @@
+##
+## Unlike `std/parseopt`, this lexer parses command line tokens as requested
+## by the caller.
+runnableExamples:
+  import std/options
+
+  type
+    Opts = object
+      paths: seq[string]
+      color: bool
+      output: string
+
+  var lexer = initCmdLexer(["--color", "/", "-o", "outfile.txt"])
+  var (kind, option) = lexer.next()
+  var opts = Opts()
+  while kind != cmdEnd:
+    if (kind == cmdLong and option == "color") or (kind == cmdShort and option == "c"):
+      opts.color = true
+    elif (kind == cmdLong and option == "output") or (kind == cmdShort and option == "o"):
+      let value = lexer.value()
+      if value.isNone:
+        quit "error: '--output' must be followed by a value"
+      opts.output = value.get()
+    elif kind == cmdValue:
+      opts.paths.add option
+
+    (kind, option) = lexer.next()
+
+  doAssert opts.paths == ["/"]
+  doAssert opts.color == true
+  doAssert opts.output == "outfile.txt"
+
+const ValueDelims* = {':', '='}
+  ## Characters that delimit between the option and its value
+
+type
+  CmdlineKind* = enum
+    cmdEnd ## End of command line
+    cmdLong ## A long option (ie. `--opt`)
+    cmdShort ## A short option (ie. `-o`)
+    cmdValue ## A value that is not an option
+
+  CmdLexer* = object
+    cmdline: seq[string] ## The command line to parse
+    index: int ## The index of `cmdline` to be processed
+    valueIdx: int ## \
+      ## For long options, the index of `ValueDelims`.
+      ##
+      ## For short options, the index of the next flag to be processed.
+      ##
+      ## Otherwise, it is unused.
+
+  LexError* = object of CatchableError
+    ## Errors occurred while lexing
+
+  UnexpectedValueError* = object of LexError
+    ## A value was provided but was not consumed
+    opt*: string ## The option that a value was provided to
+    value*: string ## The provided value
+
+proc initCmdLexer*(args: sink seq[string]): CmdLexer =
+  ## Creates a new `CmdLexer`. `args` should be the command line parameters as
+  ## returned by `os.commandLineArgs()`.
+  CmdLexer(cmdline: args)
+
+proc initCmdLexer*(args: openArray[string]): CmdLexer =
+  ## Creates a new `CmdLexer`. `args` should be the command line parameters as
+  ## returned by `os.commandLineArgs()`.
+  CmdLexer(cmdline: @args)
+
+template current(l: CmdLexer): string =
+  l.cmdline[l.index]
+
+proc isLongOpt*(s: string): bool = s.startsWith("--")
+  ## Returns whether `s` can be parsed as a long option.
+proc isShortOpt*(s: string): bool = s.startsWith('-')
+  ## Returns whether `s` can be parsed as a short option.
+proc isDash(s: string): bool = s == "-"
+proc isDashDash(s: string): bool = s == "--"
+
+proc prefix*(kind: CmdlineKind): string =
+  ## Return the prefix string for a given command line parameter kind
+  case kind
+  of cmdLong:
+    "--"
+  of cmdShort:
+    "-"
+  else:
+    ""
+
+func newUnexpectedValueError(opt, value: string): ref UnexpectedValueError =
```

Comment:
```suggestion
func newUnexpectedValueError(opt, value: sink string): ref UnexpectedValueError =
```
Not critical, given the function's exceptional nature, but it does get rid of some copies.